### PR TITLE
Improve performance of storage `S3` with large number of small files

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improved latency of reading from storage `S3` and table function `s3` with large number of small files. Now settings `remote_filesystem_read_method` and `remote_filesystem_read_prefetch` take effect while reading from storage `S3`.

Improvements:
- Read buffer for next file in stream is prepared asynchronously and is prefetched.
- Removed extra calls of `HeadObject`.
- `ListObjects` is done asynchronously now. `ListObjects` can return at most 1000 objects. So, we had several calls, which block execution of query.

### CODE REVIEW

**Comment by CurtizJ** [comment]
Tested manually in S3 with ClickHouse running in the same region as S3 bucket is placed in.

Query:
```sql
SELECT
    count(),
    uniq(_file)
FROM s3('https://s3.eu-west-1.amazonaws.com/.../*', 'CSV')

Query id: 849f7c3d-f429-427c-a779-2eb9781d77e9

┌─count()─┬─uniq(_file)─┐
│    7343 │        7343 │
└─────────┴─────────────┘

1 row in set. Elapsed: 3.219 sec. Processed 7.34 thousand rows, 316.89 KB (2.28 thousand rows/s., 98.45 KB/s.)
```

this PR:
```
clickhouse benchmark <<< "select count(), uniq(_file) from s3('https://s3.eu-west-1.amazonaws.com/.../*', 'CSV')" --cumulative

Queries executed: 10.

localhost:9000, queries 10, QPS: 0.308, RPS: 2265.240, MiB/s: 0.093, result RPS: 0.308, result MiB/s: 0.000.

0.000%          2.958 sec.
10.000%         2.962 sec.
20.000%         3.054 sec.
30.000%         3.110 sec.
40.000%         3.254 sec.
50.000%         3.291 sec.
60.000%         3.291 sec.
70.000%         3.295 sec.
80.000%         3.304 sec.
90.000%         3.549 sec.
95.000%         3.639 sec.
99.000%         3.639 sec.
99.900%         3.639 sec.
99.990%         3.639 sec.
```

master:
```
clickhouse benchmark <<< "select count(), uniq(_file) from s3('https://s3.eu-west-1.amazonaws.com/.../*', 'CSV')" --cumulative

localhost:9000, queries 10, QPS: 0.080, RPS: 583.849, MiB/s: 0.024, result RPS: 0.080, result MiB/s: 0.000.

0.000%          11.350 sec.
10.000%         11.842 sec.
20.000%         11.944 sec.
30.000%         11.974 sec.
40.000%         12.505 sec.
50.000%         12.682 sec.
60.000%         12.682 sec.
70.000%         12.859 sec.
80.000%         13.309 sec.
90.000%         13.626 sec.
95.000%         13.678 sec.
99.000%         13.678 sec.
99.900%         13.678 sec.
99.990%         13.678 sec.

File: src/Storages/StorageS3.cpp [hunk 1]
```diff
@@ -89,9 +91,23 @@ namespace ErrorCodes
...
 {
+    auto pos = path.find_last_of('/');
+    assert(pos != std::string::npos);
+
+    auto file = path.substr(pos + 1);
+    if (block.has("_path"))
+        block.getByName("_path").column->assumeMutableRef().insert(path);
 
+    if (block.has("_file"))
+        block.getByName("_file").column->assumeMutableRef().insert(file);
```
[the remark this answers is not in the document]
Comment (CurtizJ) [reply src/Storages/StorageS3.cpp:104]: Yes, it's better.

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -168,99 +178,147 @@ class StorageS3Source::DisclosedGlobIterator::Impl : WithContext
...
+    KeysWithInfo::iterator buffer_iter;
+
     Aws::S3::S3Client client;
     S3::URI globbed_uri;
     ASTPtr query;
     Block virtual_header;
-    ASTPtr filter_ast;
-    Aws::S3::Model::ListObjectsV2Request request;
-    Aws::S3::Model::ListObjectsV2Outcome outcome;
+    std::optional<ASTPtr> filter_ast;
```
[the remark this answers is not in the document]
Comment (CurtizJ) [reply src/Storages/StorageS3.cpp:309]: Yes, I can add a bool flag instead.

**Review by kssenii — APPROVED** [review]

**Comment by CurtizJ** [comment]
Performance Comparison is broken in master.

**Pull request merged**